enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -3724,6 +3726,117 @@ class AdjointGenerator
       return;
     }
 
+    if (funcName == "cblas_ddot") {
+      Type *castvals[2] = {call.getArgOperand(1)->getType(),
+                           call.getArgOperand(3)->getType()};
+      auto tupletype = StructType::get(call.getContext(), ArrayRef(castvals));
+      Value *tupleholder = BuilderZ.CreateAlloca(tupletype);
```

Comment:
Rather than making an allocation here and gep/storing into it, can you initialize to an undef then use an insertvalue?

Similarly it may be wise to first determine which values need caching and then only caching those values. For example. in ddot(a, b) you only need to cache a if b is active (since if b is inactive you shouldn't update the shadow of b, since its inactive [and otherwise may get incorrect behavior]). Additionally if a isn't overwritten between the forward and reverse (even if b is active), you don't need to cache it. The (poorly named) variable uncacheable_args_map should contain a map of arguments to whether or not the memory is overwritten.

Reply:
I might also write this to handle either ddot for doubles or sdot for floats in the same call

Reply:
> For example. in ddot(a, b) you only need to cache a if b is active (since if b is inactive you shouldn't update the shadow of b, since its inactive [and otherwise may get incorrect behavior]).

I am trying to understand this - in `ddot(a, b)`, there should never be a situation where either of them is inactive right? Unless you mean that `a` is inactive for `d/d_a ddot(a,b)` (for the cases where adjoint doesn't depend on `a`), and `b` is inactive for `d/d_b ddot(a,b)` (for the cases where adjoint doesn't depend on `b`).

Reply:
It's possible for one of the two inputs to be inactive.

For example consider

```
double sum(double* a) {
  double b[3] = {1, 1, 1};
  return ddot(a, b);
}
```

Here b is inactive, even though a is active. As such we don't need to cache a since we don't need to (and in fact would be illegal to) compute the derivative of b.